File: 02.Binary二进制/leetcode_1680_concatenatedBinary.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int concatenatedBinary(int n) {
        long long res = 0;
        int pos = 0;
        int modNum = 1e9+7;
        for (int i=1; i<=n; i++){
            if ((i&(-i)) == i){
                pos++;
            }
            res = (res<<pos | i)%modNum;
        }
        return res;
    }
};

int main(){
    Solution s;
    int n = 3;
    cout << s.concatenatedBinary(n) << endl;
}

File: 02.Binary二进制/leetcode_1680_concatenatedBinary.py
maxNum = 100_001
modNum = 1_000_000_007
preSum = [0]*maxNum
curPos = 0
for i in range(1, maxNum):
    if (i&(-i)) == i:
        curPos += 1
    preSum[i] = ((preSum[i-1]<<curPos) | i)%modNum

class Solution:
    def concatenatedBinary(self, n: int) -> int:
        return preSum[n]

if __name__ == "__main__":
    s = Solution()
    n = 3
    print(s.concatenatedBinary(n))
